### DISCUSSION TITLE
how to understand name a record directly with name of a tuple?

### DISCUSSION THREAD

**Discussion by ZhenxuanZhang-Jensen** [scikit-hep/awkward#1473] [Q&A] [answered]
Hi, experts
  I meet a code like this `name = ("a","b")`, `events[name] = ak.Array([*])`, I don't understand what this kind of expression means, how to save a tuple as a name in an awkward array? what is this mean? thanks for the help.

**Comment by agoose77** [comment, marked as answer]
`tuple` keys in NumPy and Awkward usually refer to dimensions, e.g.
```python
>>> import numpy as np
>>> matrix = np.identity(3)
>>> matrix[(0, 1)]
0
```

Awkward Array extends this to support another kind of structural feature: fields. Awkward Array let's you nest fields, e.g.

```python
ak.count(x.y.z)
```

and this is the same as writing
```python
ak.count(x[('y', 'z')])
```

Due to how indexing works, it's also the same if you didn't write any parentheses 
```python
ak.count(x['y', 'z'])
```
as Python constructs a tuple for the `__getitem__` call.

So, to answer your question, `events[("a", "b")] = ...` is setting the field `b` of `events.a` (the `RecordArray` at field `a`) to a given value. You might think that you can just write
```python
events.a.b = ...
```

But due to how Awkward Array works, this won't actually update `events`, because `events.a` is not a reference to the field `a` of `events`, but rather a shallow copy.

**Reply by ZhenxuanZhang-Jensen** [reply]
okay, cool, this is very clear, thanks a lot

**Reply by agoose77** [reply]
I should add - the order of fields and dimensions in the index is interchangeable, see https://awkward-array.readthedocs.io/en/latest/_auto/ak.Array.html#projection
